hera_qm/tests/test_xrfi.py

```diff
@@ -54,45 +64,90 @@ def setUp(self):
         raise unittest.SkipTest  # setUp has to be overridden to actually run a test
     rfi_gen = None  # Need to override this for each TestCase, usually in setUp
 
-    def _run_test(self, func, correct_flag, false_positive, nsig=4):
-        for data, rfi in self.rfi_gen():
-            f = func(data)
+    def _run_test(self, func, arg, correct_flag, false_positive, nsig=5, fmode=False):
+        for data, rfi in self.rfi_gen():      
+            f = func(data,*arg)
             if VERBOSE:
                 print self.__class__, func.__name__
-            # plot_waterfall(data, f)
+            #plot_waterfall(data, f)
             f = np.where(f > nsig, 1, 0)
             cf, fp = get_accuracy(f, rfi)
-            # plot_result(f, rfi)
-            self.assertGreater(cf, correct_flag)
-            self.assertLess(fp, false_positive)
+            print cf, fp
+            #plot_result(f, rfi)
+            if fmode:
+                print 'In failure mode now.'
+                try:
+                    self.assertLess(cf, correct_flag)
+                except AssertionError:
+                    self.assertGreater(fp, false_positive)
+            else:
+                self.assertGreater(cf, correct_flag)
+                self.assertLess(fp, false_positive)
     ans = {
-        'detrend_deriv': (.9, .1),
-        'detrend_medfilt': (.99, .01),
-        'detrend_medminfilt': (.97, .05),
-        'xrfi_simple': (.99, .1),
+        'detrend_deriv': [(.9, .6, .6),(.1,.005,.005)],
+        'detrend_medfilt': [(.9, .9, .9),(.01,.01,.01)],
+        'detrend_medminfilt': [(.97, .95, .95),(.5,.5,.5)],
+        'xrfi_simple': [(.99, .99),(.01, .01)],
         'xrfi': (.99, .01),
+        'watershed': [(.9,.9),(.1,.1)],
+    }
+    mode = {
+        'detrend_deriv': [False,False,False],
+        'detrend_medfilt': [False,False,False],
+        'detrend_medminfilt': [False,False,False],
+        'xrfi_simple': [False,False],
+        'xrfi': False,
+        'watershed': [False,True],
     }
 
+
     def test_detrend_deriv(self):
         cf, fp = self.ans['detrend_deriv']
-        self._run_test(xrfi.detrend_deriv, cf, fp, nsig=4)
+        args = [(True,True),(True,False),(False,True)]
+        mode = self.mode['detrend_deriv']
+        for i in range(3):
+            print args[i]
+            print cf[i],fp[i]
+            self._run_test(xrfi.detrend_deriv, args[i],  cf[i], fp[i], nsig=4, fmode=mode[i])
 
     def test_detrend_medfilt(self):
         cf, fp = self.ans['detrend_medfilt']
-        self._run_test(xrfi.detrend_medfilt, cf, fp, nsig=4)
+        argsList = [(8,8),(7,9),(9,7)]
```

Comment:
Something for the future - we should think about how to get the argLists and modes and ans's more closely coupled in the code. Right now there's a bit of shifting back and forth to try to understand what is being run and why.